store: new API ApplyStagedLayer

giuseppe opened the pull request:
Add a race-condition-free alternative to using CreateLayer and ApplyDiffFromStagingDirectory, ensuring the store is locked for the entire duration while the layer is being created and populated.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>


The relative patch for c/image is:

```diff
diff --git a/storage/storage_dest.go b/storage/storage_dest.go
index 88e492b7..51b30a64 100644
--- a/storage/storage_dest.go
+++ b/storage/storage_dest.go
@@ -149,7 +149,7 @@ func (s *storageImageDestination) Close() error {
 	}
 	for _, v := range s.diffOutputs {
 		if v.Target != "" {
-			_ = s.imageRef.transport.store.CleanupStagingDirectory(v.Target)
+			_ = s.imageRef.transport.store.CleanupStagedLayer(v)
 		}
 	}
 	return os.RemoveAll(s.directory)
@@ -669,11 +669,6 @@ func (s *storageImageDestination) commitLayer(index int, info addedLayerInfo, si
 			return false, fmt.Errorf("index %d out of range for configOCI.RootFS.DiffIDs", index)
 		}
 
-		layer, err := s.imageRef.transport.store.CreateLayer(id, parentLayer, nil, "", false, nil)
-		if err != nil {
-			return false, err
-		}
-
 		// let the storage layer know what was the original uncompressed layer.
 		flags := make(map[string]interface{})
 		flags[expectedLayerDiffIDFlag] = configOCI.RootFS.DiffIDs[index]
@@ -682,8 +677,15 @@ func (s *storageImageDestination) commitLayer(index int, info addedLayerInfo, si
 			Flags: flags,
 		}
 
-		if err := s.imageRef.transport.store.ApplyDiffFromStagingDirectory(layer.ID, diffOutput.Target, diffOutput, options); err != nil {
-			_ = s.imageRef.transport.store.Delete(layer.ID)
+		args := storage.ApplyStagedLayerOptions{
+			ID:          id,
+			ParentLayer: parentLayer,
+
+			DiffOutput:       diffOutput,
+			DiffOptions:      options,
+		}
+		layer, err := s.imageRef.transport.store.ApplyStagedLayer(args)
+		if err != nil {
 			return false, err
 		}

```

mtrmac commented on store.go:
```diff
@@ -71,6 +71,20 @@ type metadataStore interface {
+	Writeable    bool
+	LayerOptions *LayerOptions
+
+	StagingDirectory string
```
Would `StagingDirectory` ever differ from `DiffOutput.Target`? If not, can this be handled internally without callers having to care (and without callers possibly breaking this)?

OTOH `CleanupStagingDirectory` forces c/image to know about `.Target` anyway…

mtrmac commented on store.go:
```diff
@@ -2953,6 +2977,32 @@ func (s *store) ApplyDiffFromStagingDirectory(to, stagingDirectory string, diffO
+	}
+
+	if err := rlstore.ApplyDiffFromStagingDirectory(layer.ID, args.StagingDirectory, args.DiffOutput, args.DiffOptions); err != nil {
+		rlstore.Delete(layer.ID)
```
Please either return both errors or add a warning-level log.

Otherwise, logs from production where some layers are missing for an unknown reason are very difficult to debug. (Sure, ideally the caller should provide a unique ID that is not reused in the meantime… but if we are looking at logs, there is a bug involved. And, yes, c/image didn’t have that.)

mtrmac commented on store.go:
```diff
@@ -320,6 +334,11 @@ type Store interface {
 	// ApplyDiffFromStagingDirectory uses stagingDirectory to create the diff.
 	ApplyDiffFromStagingDirectory(to, stagingDirectory string, diffOutput *drivers.DriverWithDifferOutput, options *drivers.ApplyDiffWithDifferOpts) error
```
Should this be deprecated? I don’t have a strong opinion.

`PutLayer` exists and `CreateLayer` isn’t deprecated, that is one argument against deprecating.

giuseppe replied:
yes good idea, let's deprecate it and remove once c/image uses the new API

mtrmac commented on store.go:
```diff
@@ -71,6 +71,20 @@ type metadataStore interface {
 	rwMetadataStore
 }
 
+// PutLayerFromStagingDirectoryOptions contains options to pass to PutLayerFromStagingDirectory
```
(Documenting which of the options are mandatory and which are optional would be nice.)

mtrmac commented:
Thanks for working on this!

giuseppe commented:
@mtrmac thanks for the review.

I've addressed your comments, except one pending question: https://github.com/containers/storage/pull/1826#discussion_r1486741274

mtrmac reviewed (commented):
Changing the API, after looking a bit more, seems to to require a bunch of fairly intrusive changes all at once… so I don’t think it is worth blocking this PR on them.

Or maybe you can see some way to make all of that simple.

rhatdan commented:
LGTM

mtrmac commented on store.go:
```diff
@@ -318,11 +331,21 @@ type Store interface {
 	ApplyDiffWithDiffer(to string, options *drivers.ApplyDiffWithDifferOpts, differ drivers.Differ) (*drivers.DriverWithDifferOutput, error)
 
 	// ApplyDiffFromStagingDirectory uses stagingDirectory to create the diff.
+	// Deprecated: it will be removed soon.
```
```suggestion
	// Deprecated: it will be removed soon. Use CleanupStagingDirectory instead.
```
... or however it is going to be called.

mtrmac replied:
“`Use ApplyStagedLayer`…” here, my mistake.

mtrmac commented on store.go:
```diff
@@ -318,11 +331,21 @@ type Store interface {
 
 	// CleanupStagingDirectory cleanups the staging directory.  It can be used to cleanup the staging directory on errors
+	// Deprecated: it will be removed soon.
 	CleanupStagingDirectory(stagingDirectory string) error
```
Non-blocking: Maybe `CleanupStagedLayer`?

mtrmac reviewed (changes requested):
On second thought, I’m afraid this isn’t sufficient.

It is fine for a running system: it prevents other processes from observing the WIP layer.

But it doesn’t handle crashes sufficiently. For that, the layer metadata needs to be saved to disk with `incompleteFlag` (so that if we are recovering from a crash, we delete everything), and after the contents are set up, the flag is removed again.

(Or, hypothetically, we could first write the on-disk contents and only afterwards write the layer metadata?? But that would be a new unproven code path, and we would have to worry about re-creating a layer on top of previously partially-created files. Seems risky, when the other path is well-understood.)

Very roughly speaking, I think this can be done by applying the staged data from inside `layerStore.create`, around the place where `applyDiffWithOptions` is called for non-chunked layers.

mtrmac commented on store.go:
```diff
@@ -318,11 +331,21 @@ type Store interface {
 	ApplyDiffFromStagingDirectory(to, stagingDirectory string, diffOutput *drivers.DriverWithDifferOutput, options *drivers.ApplyDiffWithDifferOpts) error
 
 	// CleanupStagingDirectory cleanups the staging directory.  It can be used to cleanup the staging directory on errors
+	// Deprecated: it will be removed soon.
```
```suggestion
	// Deprecated: it will be removed soon. Use CleanupStagingDirectory instead.
```
... or however it is going to be called.

mtrmac commented:
The API design LGTM.

For the record:
> The relative patch for c/image is:
> ```diff
> diff --git a/storage/storage_dest.go b/storage/storage_dest.go
> index 88e492b7..51b30a64 100644
> --- a/storage/storage_dest.go
> +++ b/storage/storage_dest.go
> 
> +		layer, err := s.imageRef.transport.store.ApplyStagedLayer(args)
> +		if err != nil {
This path in c/image also needs to check for, and succeed with, `ErrDuplicateID`. And it would be convenient, and consistent with `PutLayer`, for the new function to return a `layer` value with the duplicate object in this case.

giuseppe commented:
> On second thought, I’m afraid this isn’t sufficient.
> 
> It is fine for a running system: it prevents other processes from observing the WIP layer.
> 
> But it doesn’t handle crashes sufficiently. For that, the layer metadata needs to be saved to disk with `incompleteFlag` (so that if we are recovering from a crash, we delete everything), and after the contents are set up, the flag is removed again.
> 
> (Or, hypothetically, we could first write the on-disk contents and only afterwards write the layer metadata?? But that would be a new unproven code path, and we would have to worry about re-creating a layer on top of previously partially-created files. Seems risky, when the other path is well-understood.)
> 
> Very roughly speaking, I think this can be done by applying the staged data from inside `layerStore.create`, around the place where `applyDiffWithOptions` is called for non-chunked layers.

you are right, we need to use the incompleteFlag.  I've pushed a new version where I set the incompleteFlag, barely tested.  I'll work more on it through the day

giuseppe commented:
@mtrmac what do you think of the last version?

mtrmac reviewed (commented):
I’d rather prefer if the `incompleteFlag` remained an internal implementation detail of `layers.go`.

giuseppe commented:
> I’d rather prefer if the `incompleteFlag` remained an internal implementation detail of `layers.go`.

moved the `incompleteFlag` usage to `layers.go`

mtrmac commented on layers.go:
```diff
@@ -1442,10 +1442,12 @@ func (r *layerStore) create(id string, parentLayer *Layer, names []string, mount
-	if err = r.saveFor(layer); err != nil {
-		cleanupFailureContext = "saving finished layer metadata"
-		return nil, -1, err
+	if !partial {
```
If `partial`, the code below `applyDiffWithOptions() would have updated r.bycompressedsum` modifies `layer` data and we wouldn’t save it here.

I think that might be correct enough right now (because those fields aren’t set) but it’s hard for me to track such indirect effects — the code might modify more fields in the future.

mtrmac reviewed (commented):
This is a half-way step, but `the `partial` option is still basically “leave the layer incomplete” and/or “I promise to call `applyDiffFromStagingDirectory` later” (with neither effect documented), with `store.go` being responsible for that.

It seems to me
```patch
	if diff != nil {
		if size, err = r.applyDiffWithOptions …
+	else if staged != nil {
+		if size, err = r.applyDiffFromStagingDirectory …
	else { …
```

should be possible and and not _too_ invasive a change.

giuseppe commented:
pushed a new version, I moved the `applyDiffFromStagingDirectory()` call inside `create()`

mtrmac commented on layers.go:
```diff
@@ -2482,6 +2494,8 @@ func (r *layerStore) ApplyDiffFromStagingDirectory(id, stagingDirectory string,
 		}
 	}
+
+	delete(layer.Flags, incompleteFlag)
```
This can now be dropped.

mtrmac commented on store.go:
```diff
@@ -1421,20 +1444,8 @@ func (s *store) canUseShifting(uidmap, gidmap []idtools.IDMap) bool {
 }
 
-func (s *store) PutLayer(id, parent string, names []string, mountLabel string, writeable bool, lOptions *LayerOptions, diff io.Reader) (*Layer, int64, error) {
+func (s *store) putLayerLocked(rlstore rwLayerStore, rlstores []roLayerStore, id, parent string, names []string, mountLabel string, writeable bool, lOptions *LayerOptions, diff io.Reader, slo *stagedLayerOptions) (*Layer, int64, error) {
```
Absolutely non-blocking: This could be `putLayer(… diff, slo)` which also does the locking; then `PutLayer` and `ApplyStagedLayer` would be very trivial wrappers. This works fine as it is.

giuseppe replied:
that is better, amended in the last version

mtrmac reviewed (commented):
ACK.

mtrmac reviewed (commented):
/lgtm

Thanks!

giuseppe commented:
@mtrmac the related change in c/image: https://github.com/containers/image/pull/2301
